### DISCUSSION TITLE
Smart vs dumb components

### DISCUSSION THREAD

**Discussion by benediktschlager** [pmndrs/jotai#2285] [Q&A] [answered]
Hello,
I like to ask if this is the suggested solution to our problem.
We have a TODO List component that uses an array of Todo items and creates a Todo Component for every one of these.
Now I'm wondering what the best approach is to remove a Todo item from the list when checking the Todo Item?
It basically boils down to the question if you suggest having only some smart components or having every component as a smart component is fine.
Possible solutions:
- Pass the removeTodo event handler by props
- Use the removeTodoAtom directly in the Todo Component

```typescript
type Todo = {
  name: string;
};

const todosAtom = atom<Todo[]>([{ name: "Clean up" }, { name: "Wash dishes" }]);
const removeTodoAtom = atom(undefined, (get, set, value: Todo) => {
  const todos = get(todosAtom);
  set(
    todosAtom,
    todos.filter((t) => t !== value),
  );
});

const TodoComp = ({
  name,
  removeProp,
}: {
  name: string;
  removeProp: () => void;
}) => {
  const remove = () => {
    // How to remove the todo item?
    // Use passed removeProp?
    removeProp();
    // Or turn TodoComp from a dumb component into a smart one and just use the removeTodoAtom event?
  };

  return (
    <>
      <input
        type="checkbox"
        onClick={(c) => (c ? remove() : undefined)}
      ></input>
      <label>{name}</label>
    </>
  );
};

const TodoList = () => {
  const todos = useAtomValue(todosAtom);
  const removeTodo = useSetAtom(removeTodoAtom);
  return (
    <>
      {todos.map((t) => (
        <TodoComp {...t} removeProp={() => removeTodo(t)} />
      ))}
    </>
  );
};

```

Thanks for your help!

**Comment by dai-shi** [comment, marked as answer]
I think both approaches work.
Let's wait and see if someone has an idea/feedback/experience.

**Comment by rothsandro** [comment]
The `TodoComp` already receives the todo itself as properties and doesn't access the atom for reading the state. Then I would also not update the state in the component directly but pass the event handler as prop. This makes `TodoComp` reusable and easy to test.

But I agree with @dai-shi, both approaches work.
